### PR TITLE
Add support for Dart-side `group()`s

### LINKED ISSUE #1619: Add support for Dart-side `group()`s and multiple `test()`s
A subtask of #1501.

### PR DESCRIPTION
**Pull request by bartekpacia**
This PR resolves #1619.

### CODE REVIEW

**Comment by bartekpacia** [comment]
Hot Restart status: works, but per-file, instead of per-`test()`.

**Comment by mkhtradm01** [comment]
We really need this feature, looking forward to have it release 🙌

File: packages/patrol/example/ios/RunnerTests/RunnerTests.swift [hunk 1]
```diff
@@ -39,4 +37,12 @@ final class RunnerTests: XCTestCase {
     let result = PatrolUtils.createMethodName(fromPatrolGeneratedGroup: input)
     XCTAssertEqual(result, expectedOutput)
   }
+
+  func testCreateMethodNameFromPatrolGeneratedGroup_ExampleWithGroups() {
+    let input = "example_test at the beginning"
+    let expectedOutput = "exampleTestatthebeginning"  // FIXME: This is temporary
+    let result = PatrolUtils.createMethodName(fromPatrolGeneratedGroup: input)
+    NSLog(result)
```
Comment (bartekpacia) [inline packages/patrol/example/ios/RunnerTests/RunnerTests.swift:45]: It's OK for this FIXME to land in a `dev` release on `develop` branch.
I'll get to it once we'll be stabilizing before merging to `master`.

File: .github/workflows/patrol-prepare.yaml [hunk 2]
```diff
@@ -155,13 +155,13 @@ jobs:
       - name: Run unit tests
         working-directory: packages/patrol/example/ios
         run: |
-          xcodebuild test \
+          set -o pipefail && xcodebuild test \
```
Comment (shilangyu) [inline .github/workflows/patrol-prepare.yaml:158]: any reason it is not a separate line?

File: packages/patrol/lib/src/extensions.dart [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+/// Provides convenience methods for [Invoker].
+@internal
+extension InvokerX on Invoker {
+  /// Returns the full name of the current test (names of all ancestor groups +
+  /// name of the current test).
+  String fullCurrentTestName() {
+    final parentGroupName = liveTest.groups.last.name;
+    final testName = liveTest.individualName;
+
+    return '$parentGroupName $testName';
```
Comment (shilangyu) [inline packages/patrol/lib/src/extensions.dart:14]: > names of all ancestor groups

Is that true? You seem to be only collecting the last one.

File: packages/patrol/lib/src/extensions.dart [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+// ignore: implementation_imports
+import 'package:test_api/src/backend/invoker.dart';
+
+/// Provides convenience methods for [Invoker].
+@internal
+extension InvokerX on Invoker {
+  /// Returns the full name of the current test (names of all ancestor groups +
+  /// name of the current test).
+  String fullCurrentTestName() {
+    final parentGroupName = liveTest.groups.last.name;
```
Comment (shilangyu) [inline packages/patrol/lib/src/extensions.dart:11]: What if there are no groups?

File: packages/patrol/lib/src/extensions.dart (hunk 3 above)
Comment (shilangyu) [inline packages/patrol/lib/src/extensions.dart:14]: Is there any requirement that this should be unique in a test suite?

File: packages/patrol/example/integration_test/open_app_test.dart [hunk 5]
```diff
@@ -5,14 +5,14 @@ import 'package:flutter/material.dart';
...
-  }
-
-  patrol('counter state is the same after switching apps', ($) async {
+  patrol('open maps', ($) async {
+    late String mapsId;
+    if (Platform.isIOS) {
+      mapsId = 'com.apple.Maps';
+    } else if (Platform.isAndroid) {
+      mapsId = 'com.google.android.apps.maps';
+    }
```
Comment (shilangyu) [inline packages/patrol/example/integration_test/open_app_test.dart:14]: What happens in the else case? mapsId is not initialized?

File: packages/patrol/ios/Classes/PatrolAppServiceClient.swift [hunk 6]
```diff
@@ -55,3 +57,36 @@ import NIO
...
   }
 }
+
+extension Patrol_DartGroupEntry {
+
+  func listTestsFlat(parentGroupName: String) -> [Patrol_DartGroupEntry] {
+    var tests = [Patrol_DartGroupEntry]()
+
+    for test in self.entries {
+      var test = test
```
Comment (shilangyu) [inline packages/patrol/ios/Classes/PatrolAppServiceClient.swift:67]: i might be missing something, please explain this to me

File: packages/patrol/example/integration_test/open_app_test.dart (hunk 5 above)
Comment (bartekpacia) [reply packages/patrol/example/integration_test/open_app_test.dart:14]: If we get to `else` it means it's neither Android nor iOS, so we don't care.

File: packages/patrol/ios/Classes/PatrolAppServiceClient.swift (hunk 6 above)
Comment (bartekpacia) [reply packages/patrol/ios/Classes/PatrolAppServiceClient.swift:67]: `test` defined in for-loop body is a `let`, I redeclare it as `var` so I can modify (actually: copy-on-write) its `name` a few lines below

File: packages/patrol/ios/Classes/PatrolAppServiceClient.swift [hunk 7]
```diff
@@ -55,3 +57,36 @@ import NIO
...
+
+      if test.type == Patrol_DartGroupEntry.GroupEntryType.test {
+        if parentGroupName.isEmpty {
+          // This case is invalid, because every test will have at least
+          // 1 named group - its filename.
+
+          continue  // Ignore - what else can we do?
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/patrol/ios/Classes/PatrolAppServiceClient.swift:74]: If this case is executed, this means there's a serious programming error somewhere in Patrol.

I think log message to warn about this + `continue` is good enough.

File: packages/patrol/lib/src/extensions.dart (hunk 3 above)
Comment (bartekpacia) [reply packages/patrol/lib/src/extensions.dart:14]: Yeah, but it's a requirement that's already enforced by `package:test` itself.

File: packages/patrol/ios/Classes/PatrolAppServiceClient.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (shilangyu) [reply packages/patrol/ios/Classes/PatrolAppServiceClient.swift:74]: serious programming error == panic ([fatalError](https://developer.apple.com/documentation/swift/fatalerror(_:file:line:)) in Swift land)

File: packages/patrol/example/integration_test/open_app_test.dart (hunk 5 above)
Comment (shilangyu) [reply packages/patrol/example/integration_test/open_app_test.dart:14]: Then i would recommend not marking it as late (hides the problem) and instead add an else branch where you throw some error.

File: packages/patrol/lib/src/extensions.dart (hunk 3 above)
Comment (shilangyu) [reply packages/patrol/lib/src/extensions.dart:14]: is it?

```sh
➜ cat test/test_test.dart
import 'package:test/expect.dart';
import 'package:test/scaffolding.dart';

void main() {
  test('group test', () {
    expect(true, true);
  });
  group('group', () {
    test('test', () {
      expect(true, true);
    });
  });
}

➜ dart test test/test_test.dart
00:00 +2: All tests passed!
```

Comment (bartekpacia) [reply packages/patrol/lib/src/extensions.dart:14]: you're right. thanks

We could use `_` to separate groups. It'll be required anyway if we want to reconstruct the original test suite hierarchy later.

File: packages/patrol/lib/src/extensions.dart (hunk 4 above)
Comment (bartekpacia) [reply packages/patrol/lib/src/extensions.dart:11]: (Assuming no fatal programming error in Patrol) 

Every `test()` always has at least one parent group - the one in `test_bundle.dart`:

<details><summary>sample</summary>

```dart
  // START: GENERATED TEST GROUPS
  group('android_app_test', android_app_test.main);
  group('example_test', example_test.main);
  group('open_app_test', open_app_test.main);
  group('open_quick_settings_test', open_quick_settings_test.main);
  group('permissions.notifications_test', permissions__notifications_test.main);
  group('permissions.permissions_location_test', permissions__permissions_location_test.main);
  group('permissions.permissions_many_test', permissions__permissions_many_test.main);
  group('service_airplane_mode_test', service_airplane_mode_test.main);
  group('service_bluetooth_test', service_bluetooth_test.main);
  group('service_cellular_test', service_cellular_test.main);
  group('service_dark_mode_test', service_dark_mode_test.main);
  group('service_wifi_test', service_wifi_test.main);
  group('swipe_test', swipe_test.main);
  group('webview_hackernews_test', webview_hackernews_test.main);
  group('webview_leancode_test', webview_leancode_test.main);
  group('webview_stackoverflow_test', webview_stackoverflow_test.main);
  // END: GENERATED TEST GROUPS
```

</details>

File: .github/workflows/patrol-prepare.yaml (hunk 2 above)
Comment (bartekpacia) [reply .github/workflows/patrol-prepare.yaml:158]: copied from `xcbeautify`'s docs, that's why haha

File: packages/patrol/android/src/main/kotlin/pl/leancode/patrol/ContractsExtensions.kt [hunk 8]
```diff
@@ -184,21 +185,28 @@ fun Contracts.Selector.toBySelector(): BySelector {
...
-
-    return groups
+/**
+ * Flattens the structure of a DartTestSuite into a flat list of tests.
+ */
+fun DartGroupEntry.listTestsFlat(parentGroupName: String = ""): List<DartGroupEntry> {
+    val tests = mutableListOf<DartGroupEntry>()
+
+    for (test in entriesList) {
+        if (test.type == DartGroupEntry.GroupEntryType.TEST) {
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/patrol/android/src/main/kotlin/pl/leancode/patrol/ContractsExtensions.kt:195]: thanks for the catch. unfortunately, some ugly `UNRECOGNIZED` is generated, which we don't really care about.

I think a good ol' `if` is fine here.

Screenshot 2023-09-12 at 3 03 54 PM

File: packages/patrol/example/integration_test/open_app_test.dart (hunk 5 above)
Comment (bartekpacia) [reply packages/patrol/example/integration_test/open_app_test.dart:16]: right

done

File: packages/patrol/ios/Classes/PatrolAppServiceClient.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/patrol/ios/Classes/PatrolAppServiceClient.swift]: thanks, that's a goooood idea

File: packages/patrol/lib/src/extensions.dart (hunk 3 above)
Comment (shilangyu) [reply packages/patrol/lib/src/extensions.dart:14]: Do you plan to address it now or later?

File: packages/patrol/example/integration_test/swipe_test.dart [hunk 9]
```diff
@@ -8,6 +8,8 @@ void main() {
     appId = 'com.apple.Preferences';
   } else if (Platform.isAndroid) {
     appId = 'com.android.settings';
+  } else {
+    throw UnsupportedError('Unsupported platform');
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/patrol/example/integration_test/swipe_test.dart:12]: thx

File: packages/patrol/example/integration_test/open_app_test.dart [hunk 10]
```diff
@@ -5,14 +5,16 @@ import 'package:flutter/material.dart';
...
-  late String mapsId;
-  if (Platform.isIOS) {
-    mapsId = 'com.apple.Maps';
-  } else if (Platform.isAndroid) {
-    mapsId = 'com.google.android.apps.maps';
-  }
-
-  patrol('counter state is the same after switching apps', ($) async {
+  patrol('open maps', ($) async {
+    late String mapsId;
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/patrol/example/integration_test/open_app_test.dart]: thx:)

File: packages/patrol/example/integration_test/open_app_test.dart [hunk 11]
```diff
@@ -26,4 +28,28 @@ void main() {
 
     expect($(#counterText).text, '1');
   });
+
+  patrol('open browser', ($) async {
+    late String browserId;
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/patrol/example/integration_test/open_app_test.dart]: fixed

File: packages/patrol_cli/lib/src/test_bundler.dart [hunk 12]
```diff
@@ -71,17 +71,21 @@ Future<void> main() async {
...
   // This test must be the first to run. If not, the native side likely won't
   // receive any tests, and everything will fall apart.
   test('patrol_test_explorer', () {
+    // Maybe somewhat counterintuitively, this callback runs *after* the calls
+    // to group() below.
     final topLevelGroup = Invoker.current!.liveTest.groups.first;
     final dartTestGroup = createDartTestGroup(topLevelGroup);
     testExplorationCompleter.complete(dartTestGroup);
+    print('PATROL_DEBUG: group structure:');
+    printGroupStructure(dartTestGroup);
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/patrol_cli/lib/src/test_bundler.dart]: Yup, I'd say it's quite useful. Changed it to:

```dart
print('patrol_test_explorer: obtained Dart-side test hierarchy:');
```

[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/patrol_cli/lib/src/test_bundler.dart]: Can be always removed later.

File: packages/patrol/lib/src/extensions.dart (hunk 3 above)
Comment (bartekpacia) [reply packages/patrol/lib/src/extensions.dart:14]: Later, as part of #1502.

File: packages/patrol/example/integration_test/example_test.dart [hunk 13]
```diff
@@ -1,27 +1,47 @@
...
+import 'package:test_api/src/backend/invoker.dart';
 
 import 'common.dart';
 
 void main() {
-  patrol(
-    'counter state is the same after going to Home and switching apps',
-    ($) async {
-      await createApp($);
+  // FIXME: iOS implementation not tested with non-alphabetic test names.
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/patrol/example/integration_test/example_test.dart]: It's fixed, non-alphabetic characters are removed now when creating Objective-C method name out of the test name ([impl](https://github.com/leancodepl/patrol/blob/ef29f1de8889b94d66cb4dccbc0d2c5d87c66534/packages/patrol/ios/Classes/PatrolUtils.m#L62-L86)) ([test](https://github.com/leancodepl/patrol/blob/ef29f1de8889b94d66cb4dccbc0d2c5d87c66534/packages/patrol/example/ios/RunnerTests/RunnerTests.swift#L81-L87))

File: packages/patrol/example/integration_test/example_test.dart [hunk 14]
```diff
@@ -1,27 +1,47 @@
...
+        await _testBody($);
+      });
+      patrol('second', ($) async {
+        await _testBody($);
+      });
+    });
+  });
+}
+
+// FIXME: Only for debugging and development. To be removed.
```
[the remark this answers is not in the document]
Comment (bartekpacia) [reply packages/patrol/example/integration_test/example_test.dart]: right, sorted out

**Review by shilangyu — APPROVED** [review]

**Comment by bartekpacia** [comment]
@mkhtradm01 Check out new `dev` releases, `group()`s should work there.

**Comment by mkhtradm01** [comment]
@bartekpacia Thank you 🙌

**Pull request merged**